crates/torii/core/src/sql.rs

```diff
@@ -1175,9 +1198,383 @@ impl Sql {
+            } else {
+                ByteArray::cairo_deserialize(&name, 0)
+                    .expect("Return value not ByteArray")
+                    .to_string()
+                    .expect("Return value not String")
+            };
+
+            let symbol = provider
+                .call(
+                    FunctionCall {
+                        contract_address,
+                        entry_point_selector: get_selector_from_name("symbol").unwrap(),
+                        calldata: vec![],
+                    },
+                    BlockId::Tag(BlockTag::Pending),
+                )
+                .await?;
+            let symbol = if symbol.len() == 1 {
+                parse_cairo_short_string(&symbol[0]).unwrap()
+            } else {
+                ByteArray::cairo_deserialize(&symbol, 0)
+                    .expect("Return value not ByteArray")
+                    .to_string()
+                    .expect("Return value not String")
+            };
+
+            let decimals = provider
+                .call(
+                    FunctionCall {
+                        contract_address,
+                        entry_point_selector: get_selector_from_name("decimals").unwrap(),
+                        calldata: vec![],
+                    },
+                    BlockId::Tag(BlockTag::Pending),
+                )
+                .await?;
+            let decimals = u8::cairo_deserialize(&decimals, 0).expect("Return value not u8");
+
+            // Insert the token into the tokens table
+            self.query_queue.enqueue(
+                "INSERT INTO tokens (id, contract_address, name, symbol, decimals) VALUES (?, ?, \
+                 ?, ?, ?)",
+                vec![
+                    Argument::String(format!("{:#x}", contract_address)),
+                    Argument::FieldElement(contract_address),
+                    Argument::String(name),
+                    Argument::String(symbol),
+                    Argument::Int(decimals.into()),
+                ],
+            );
+        }
+
+        // Now proceed with the transfer handling
+        // Insert transfer event to erc20_transfers table
+        {
+            let insert_query = "INSERT INTO erc_transfers (contract_address, from_address, \
+                                to_address, amount, token_id, executed_at) VALUES (?, ?, ?, ?, ?, \
+                                ?)";
+
+            self.query_queue.enqueue(
+                insert_query,
+                vec![
+                    Argument::FieldElement(contract_address),
+                    Argument::FieldElement(from),
+                    Argument::FieldElement(to),
+                    Argument::String(u256_to_sql_string(&amount)),
+                    Argument::String(token_id.clone()),
+                    Argument::String(utc_dt_string_from_timestamp(block_timestamp)),
+                ],
+            );
+        }
+
+        // Update balances in erc20_balance table
```

Comment:
I'm not sure about the manipulation of balances here. Because until we reach head of the chain, the balance will not be accurate.

Or do we assume that we should reach the head of the chain pretty quickly, and then it's fine to keep updating the balances?

From a client perspective, let's take Controller as an example, we may want to query the current balance with regular interval when it's needed. In this case, using the balance in the DB makes little sense? Or once again we assume we should be quickly in the head of chain, and pulling data from the DB or the chain should return the same result.

Reply:
> we assume that we should reach the head of the chain pretty quickly

yeah thats the assumption we make

say we only provide the info of which tokens are owned by user for that also we would need to index all the transfer events anyways and if we are lagging behind in indexing controller might not get info about some of the token.

so even by not keeping the balance we dont fully solve that issue. and not to mention that if user has large amount of tokens controller would need to make a rpc call to get balance for each of them.